OpenRA.Mods.D2k/Traits/Sandworm.cs

```diff
@@ -99,33 +100,35 @@ void RescanForTargets(Actor self)
 				if (!actor.IsInWorld)
 					continue;
 
-				// TODO: Test if we really want to ignore actors that are on rock
 				if (!mobile.Value.CanEnterCell(actor.Location, null, false))
 					continue;
 
-				var noise = actor.TraitOrDefault<AttractsWorms>();
-				if (noise == null)
+				var noiseTraits = actor.TraitsImplementing<AttractsWorms>().Where(x => !x.IsTraitDisabled);
+				if (!noiseTraits.Any())
 					continue;
 
-				var distance = actor.CenterPosition - self.CenterPosition;
-				var length = distance.Length;
+				foreach (var noiseTrait in noiseTraits)
+				{
+					var distance = actor.CenterPosition - self.CenterPosition;
+					var length = distance.Length;
 
-				// Actor is too far to be heard
-				if (noise.Info.Range[noise.Info.Range.Length - 1].Range < length)
-					continue;
+					// Actor is too far to be heard
+					if (noiseTrait.Info.Range[noiseTrait.Info.Range.Length - 1].Range < length)
+						continue;
 
-				// If close enough, we don't care about other actors
-				if (length <= Info.IgnoreNoiseAttackRange.Range)
-				{
-					self.CancelActivity();
-					attackTrait.Value.ResolveOrder(self, new Order("Attack", actor, true) { TargetActor = actor });
-					return;
```

Comment:
I'll leave that to you.  The important part here is to move the noise calculation where it conceptually belongs: in AttractsNoise.